engines/stark/ui/menu/dialogmenu.cpp

```diff
@@ -0,0 +1,386 @@
+	uint bottom = _logFrame->getText()->getTargetHeight() + pos.y;
+	uint height;
+	DialogLineText *dialogLineText;
+
+	Diary::ConversationLog dialog = StarkDiary->getDialog(_curLogIndex);
+
+	while (_nextLineIndex < dialog.lines.size()) {
+		height = 0;
+
+		dialogLineText = new DialogLineText(_gfx, _curLogIndex, _nextLineIndex, boxWidth);
+		height = dialogLineText->getHeight();
+
+		if (pos.y + height + space > bottom) {
+			break;
+		}
+
+		dialogLineText->setPosition(pos);
+		_dialogLineTexts.push_back(dialogLineText);
+
+		pos.y += height + space;
+
+		++_nextLineIndex;
+	}
+
+	_widgets[kWidgetLogBack]->setVisible(_startLineIndex > 0);
+	_widgets[kWidgetLogNext]->setVisible(_nextLineIndex < dialog.lines.size());
+}
+
+void DialogScreen::backHandler() {
+	StarkUserInterface->backPrevScreen();
+}
+
+void DialogScreen::indexBackHandler() {
+	_nextTitleIndex = _prevTitleIndexStack.back();
+	_prevTitleIndexStack.pop_back();
+	loadIndex();
+}
+
+void DialogScreen::indexNextHandler() {
+	_prevTitleIndexStack.push_back(_startTitleIndex);
+	loadIndex();
+}
+
+void DialogScreen::logBackHandler() {
+	_nextLineIndex = _prevLineIndexStack.back();
+	_prevLineIndexStack.pop_back();
+	loadDialog();
+}
+
+void DialogScreen::backIndexHandler() {
+	freeDialogLineTexts();
+	_prevLineIndexStack.clear();
+
+	_widgets[kWidgetLogBack]->setVisible(false);
+	_widgets[kWidgetLogNext]->setVisible(false);
+	_widgets[kWidgetIndex]->setVisible(false);
+
+	_nextTitleIndex = _startTitleIndex;
+	loadIndex();
+}
+
+void DialogScreen::logNextHandler() {
+	_prevLineIndexStack.push_back(_startLineIndex);
+	loadDialog();
+}
+
+void DialogScreen::freeLogTitleWidgets() {
+	uint size = _widgets.size();
+	for (uint i = 0; i < size - _dialogTitleWidgetOffset; ++i) {
+		delete _widgets.back();
+		_widgets.pop_back();
+	}
+}
+
+void DialogScreen::freeChapterTitleTexts() {
+	for (uint i = 0; i < _chapterTitleTexts.size(); ++i) {
+		delete _chapterTitleTexts[i];
+	}
+	_chapterTitleTexts.clear();
+}
+
+void DialogScreen::freeDialogLineTexts() {
+	for (uint i = 0; i < _dialogLineTexts.size(); ++i) {
+		delete _dialogLineTexts[i];
+	}
+	_dialogLineTexts.clear();
+}
+
+ChapterTitleText::ChapterTitleText(Gfx::Driver *gfx, uint chapter) :
+		_pos(), _text(gfx) {
+	Common::String text = Common::String::format(
+			"%s: %s",
+			StarkGameChapter->getChapterTitle(chapter).c_str(),
+			StarkGameChapter->getChapterSubtitle(chapter).c_str());
+	text.toUppercase();
+
+	_text.setText(text);
+	_text.setColor(_color);
+	_text.setFont(FontProvider::kCustomFont, 5);
+}
+
+DialogLineText::DialogLineText(Gfx::Driver *gfx, uint logIndex, uint lineIndex, uint boxWidth) :
+		_namePos(), _linePos(), _boxWidth(boxWidth),
+		_nameText(gfx), _lineText(gfx) {
+	Diary::ConversationLogLine logLine = StarkDiary->getDialog(logIndex).lines[lineIndex];
+
+	Common::String name = StarkGlobal->getCharacterName(logLine.characterId);
+	name.toUppercase();
+
+	uint color = name == "APRIL" ? _textColorApril : _textColorNormal;
```

Comment:
This is how this test is done in other parts of the engine: https://github.com/residualvm/residualvm/blob/8b5e7433f5cf1a17205c1caa9b5f3788f405e45b/engines/stark/resources/speech.cpp#L131